### PR TITLE
Image Classification API: Fix processing incomplete batch(<batchSize), images processed per epoch , enable EarlyStopping without Validation Set. Fixes #4274 and #4286    

### LINKED ISSUE #4274: [Image Classification API] No evaluation when batchSize parameter > # of instances in dataset
### System information

- **OS version/distro**: Windows 10
- **.NET Version (eg., dotnet --info)**: .NET Core 2.2
- **ML.NET Version**: 1.4.0-preview

### Issue

- **What did you do?**

Tried to train an image classification model using the Image Classification API. The value set for `batchSize` parameter is 300. Meanwhile then number of data instances in the test set is 182.

- **What happened?**

No evaluation takes place. 0 batches are processed. 

- **What did you expect?**

The model to train and for it to evaluate the number of instances provided. In this case since the number of data instances is less than the amount set for the `batchSize` parameter, it would process 1 batch instead of 0.

The model to evaluate

### Source code / logs

Pipeline:

```csharp
var trainingPipeline =
                mapLabelTransform
               .Append(mlContext.Model.ImageClassification(
                   featuresColumnName: "ImagePath",
                   labelColumnName: "LabelAsKey",
                   arch: ImageClassificationEstimator.Architecture.ResnetV2101,
                   epoch: 100,
                   batchSize: 300,
                   testOnTrainSet: false,
                   metricsCallback: (metrics) => Console.WriteLine(metrics),
                   validationSet: transformedTestData,
                   reuseTrainSetBottleneckCachedValues: true,
                   reuseValidationSetBottleneckCachedValues: true));
```

Output:

```text
Number of rows 182
Phase: Training, Dataset used: Validation, Batch Processed Count:   0, Epoch:  93, Accuracy:        NaN
Phase: Training, Dataset used: Validation, Batch Processed Count:   0, Epoch:  94, Accuracy:        NaN
Phase: Training, Dataset used: Validation, Batch Processed Count:   0, Epoch:  95, Accuracy:        NaN
Phase: Training, Dataset used: Validation, Batch Processed Count:   0, Epoch:  96, Accuracy:        NaN
Phase: Training, Dataset used: Validation, Batch Processed Count:   0, Epoch:  97, Accuracy:        NaN
Phase: Training, Dataset used: Validation, Batch Processed Count:   0, Epoch:  98, Accuracy:        NaN
Phase: Training, Dataset used: Validation, Batch Processed Count:   0, Epoch:  99, Accuracy:        NaN
```

When the `batchSize` is set equal to the number of rows (in this case 182), this is the output:

```text
Phase: Training, Dataset used: Validation, Batch Processed Count:   1, Epoch:  95, Accuracy:          1
Phase: Training, Dataset used: Validation, Batch Processed Count:   1, Epoch:  96, Accuracy:          1
Phase: Training, Dataset used: Validation, Batch Processed Count:   1, Epoch:  97, Accuracy:          1
Phase: Training, Dataset used: Validation, Batch Processed Count:   1, Epoch:  98, Accuracy:          1
Phase: Training, Dataset used: Validation, Batch Processed Count:   1, Epoch:  99, Accuracy:          1
```

### PR DESCRIPTION
**Pull request by ashbhandare**
1)Previously, if the images left were not enough to for a batch of batchSize, the batch would not be processed. Fixed to process incomplete batch in training and validation.
2)There was a bug where the batchIndex was not getting reset when the last batch was incomplete(< batchSize). Fixed to reset batchIndex.
3)EarlyStopping was not triggering when validation set is not provided. Fixed.

fixes #4274 #4286

### CODE REVIEW

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 1]
```diff
@@ -429,6 +430,34 @@ private IDataView GetShuffledData(string path)
                         }
                     }
 
+                    //Process last incomplete batch
+                    if (batchIndex > 0)
+                    {
+                        runner.AddInput(new Tensor(featureBatchPtr, featureTensorShape, TF_DataType.TF_FLOAT, featureBatchSizeInBytes), 0)
+                                .AddInput(new Tensor(labelBatchPtr, labelTensorShape, TF_DataType.TF_INT64, labelBatchSizeInBytes), 1)
+                                .Run();
+
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: At this point we know the batch size is not equal to batchSize so we will need to change the shape(featureTensorShape,labelTensorShape) of the input tensor so TF does not process the entire batch and same for batchSizeInBytes.

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 2]
```diff
@@ -478,6 +515,22 @@ private IDataView GetShuffledData(string path)
...
+                        outputTensors[0].ToScalar<float>(ref accuracy);
+                        metrics.Train.Accuracy += accuracy;
+                        metrics.Train.BatchProcessedCount += 1;
+                        batchIndex = 0;
+
+                        outputTensors[0].Dispose();
+                    }
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: At this point we know the batch size is not equal to batchSize so we will need to change the shape(featureTensorShape,labelTensorShape) of the input tensor so TF does not process the entire batch and same for batchSizeInBytes.

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs (hunk 1 above)
Comment (ashbhandare) [reply src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: done.

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 3]
```diff
@@ -186,6 +186,8 @@ private void CheckTrainingParameters(ImageClassificationEstimator.Options option
 
             if (_session.graph.OperationByName(_labelTensor.name.Split(':')[0]) == null)
                 throw Host.ExceptParam(nameof(options.TensorFlowLabel), $"'{options.TensorFlowLabel}' does not exist in the model");
+            if (options.EarlyStoppingCriteria != null && options.ValidationSet == null && options.TestOnTrainSet == false)
+                throw Host.ExceptParam(nameof(options.EarlyStoppingCriteria), $"No Validation dataset provided and testing on Train disabled, Early Stopping not supported.");
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: >"No Validation dataset provided and testing on Train disabled, Early Stopping not supported. [](start = 79, length = 92)

Lets improve the exception message "Early stopping enabled but unable to find a validation set and/or train set testing disabled. Please disable early stopping or either provide a validation set or enable train set training."

File: src/Microsoft.ML.Dnn/ImageClassificationTransform.cs [hunk 4]
```diff
@@ -429,6 +432,42 @@ private IDataView GetShuffledData(string path)
                         }
                     }
 
+                    //Process last incomplete batch
+                    if (batchIndex > 0)
+                    {
+                        featureTensorShape[0] = batchIndex;
+                        featureBatchSizeInBytes = sizeof(float) * featureBatch.Length * batchIndex / batchSize;
```
Comment (codemzs) [inline src/Microsoft.ML.Dnn/ImageClassificationTransform.cs]: > sizeof(float) * featureBatch.Length * batchIndex / batchSize; [](start = 49, length = 62)

sizeof(float) * featureValues.Length * batchIndex

**Comment by codemzs** [comment]
@ashbhandare Please also remove the WIP tag.
